### PR TITLE
add proactive language reporting by using major-mode

### PR DESCRIPTION
**Pull request by Anton-Latukha**
Relates to solving the: https://github.com/wakatime/wakatime-mode/issues/37

`master` algorythm currently provides no info on language, so `wakatime` does the default - tries to autodetect the languages.

Since:
```
wakatime --help

...

  --language LANGUAGE   Optional language name. If valid, takes priority over
    auto-detected language.
```

This patch adds that if case:
```
  if `--language` passes the valid as valis
    then success
    else waketime uses default - tries to autodetect language
```

What we do is simply:
matching https://www.emacswiki.org/emacs/List_Of_Major_And_Minor_Modes#toc2
to https://github.com/wakatime/wakatime/blob/master/wakatime/languages/default.json

In other words:
Emacs does: `c++-mode -> c++ -> wakatime`
Wakatime does:  `c++ -> C++`

Profit.

### CODE REVIEW

**Comment by thomasf** [comment]
I recommend against merging this because it lacks user configuration.

In particular because of https://github.com/wakatime/wakatime-mode/issues/37#issuecomment-702188769

**Comment by Anton-Latukha** [comment]
@thomasf, please, don't be ridiculous.

Please, read the PR header and its comments in the code before blocking the PR.

**Comment by Anton-Latukha** [comment]
Multiple major modes are not supported by Emacs project.

Nobody should support your unofficial hacking system, which officially deprecated by Emacs.

It is if I would launch a decompiler, open ASM, and hack Windows and then should and complain to Microsoft and about Microsoft that they are that they do not support me and my hacks.

**Comment by Anton-Latukha** [comment]
Since in majority of cases this proactive reporting works way better then heuristic autodetection
You state:
> In several situations this will be worse than the current pygments detection.

Since I refuted one you provided, please can you show valid cases. Where actively sending info that gets checked works worse than sending no info.

**Comment by thomasf** [comment]
> @thomasf, please, don't be ridiculous.
> 
> Please, read the PR header and its comments in the code before blocking the PR.

if say js2-mode or one of the other examples with the same problem it won't work,. so you don't want to it to be able to work for , It's not a complete solution.

**Comment by Anton-Latukha** [comment]
There is no complete solutions.

**Comment by Anton-Latukha** [comment]
Ever.

**Comment by thomasf** [comment]
> > > @thomasf, please, don't be ridiculous.
> > > Please, read the PR header and its comments in the code before blocking the PR.
> > 
> > 
> > if say js2-mode or one of the other examples with the same problem it won't work,. so you don't want to it to be able to work for , It's not a complete solution.
> 
> Jesus, I already responded in [#37 (comment)](https://github.com/wakatime/wakatime-mode/issues/37#issuecomment-702070724), it would ignore the `js2` and others and would use the autodetection.

Why not just always rely on auto detection then if you don't intend this feature which easily can gain extra features to handle this to have it?

**Comment by Anton-Latukha** [comment]
Do you understand the words "`If valid, takes priority over auto-detected language.`" What part of it you do not understand?

**Comment by thomasf** [comment]
> Do you understand the words "`--language LANGUAGE Optional language name. If valid, takes priority over auto-detected language.`" What part of it you do not understand?

As I understand it this patch is trying to solve the issue where wakatime-cli fails to detect a language, if js2-mode is used and wakatime fails to detect the JS language it will still fail.

I mean it's super easy to add support for a defcustom, I have no idea why you are resisting a more robust internal detection.

**Comment by Anton-Latukha** [comment]
I do not understand why you actively trying to block the valid effective solution over some imaginary code.

After this code gets merged - anybody can take that code and form and send patches with as many `defcustom`s and functions as they want.

**Comment by Anton-Latukha** [comment]
There are such people. Workin pragmatic code provided them for free - does not satisfies them, and moreover - they complain a lot.

**Comment by thomasf** [comment]
> There is no complete solutions.

it is not complete enough to solve #37 which is the scope defined in this pull request.

**Comment by Anton-Latukha** [comment]
Once more, look into basic statement that `wakatime` makes. This PR does not change the current behaviour in case of `js2-mode` or multi-major modes (which, once more, - are not supported by Emacs).

You definitely still not read the wakatime documentation and processes, I do not know what you are talking about.

**Comment by Anton-Latukha** [comment]
Statement about `js2-mode` is a false, because `js2` is not a recognized value by `wakatime --language` option:
https://github.com/wakatime/wakatime/blob/master/wakatime/languages/default.json

And if one would take effort of reading two lines about what `--language` option is - it explains itself.

**Comment by thomasf** [comment]
https://github.com/wakatime/wakatime-mode/issues/37#issuecomment-702509731

**Comment by Anton-Latukha** [comment]
I responded to you there.

Still not applicabel example. It does not have intersection with the code and process discussed.

File: wakatime-mode.el [hunk 1]
```diff
@@ -169,12 +169,26 @@ the wakatime subprocess occurs."
...
   "Return client command executable and arguments.
    Set SAVEP to non-nil for write action."
-  (format "%s%s--file \"%s\" --plugin \"%s/%s\" --time %.2f%s%s"
+  (format "%s%s--file \"%s\" --plugin \"%s/%s\" --language \"%s\" --time %.2f%s%s"
     (if (s-blank wakatime-python-bin) "" (format "\"%s\" " wakatime-python-bin))
     (if (s-blank wakatime-cli-path) "wakatime " (format "\"%s\" " wakatime-cli-path))
     (buffer-file-name (current-buffer))
     wakatime-user-agent
     wakatime-version
+    ;; --language LANGUAGE Optional language name. If valid, takes priority over auto-detected language.
```
Comment (alanhamlett) [inline wakatime-mode.el:178]: Is the major mode always accurate? If not, we should use `--alternate-language`.

**Comment by alanhamlett** [comment]
> Statement about `js2-mode` is a false, because `js2` is not a recognized value by `wakatime --language` option

WakaTime uses `--language` without validation. Currently only Vim modelines are validated in wakatime-cli, otherwise `--language` is sent only with case changes, for ex: `c++` -> `C++`.
